File: tests/unit/helpers/fake-supabase.mjs
// Lokalny FAŁSZYWY serwer zgodny z używanym podzbiorem REST Supabase (Auth + PostgREST) — wyłącznie do testów.
// Emuluje: logowanie hasłem, odświeżanie z rotacją tokenów, wygasanie tokenów, RLS (każdy użytkownik widzi tylko swoje
// wiersze, anon nie ma dostępu), wstawianie z pominięciem duplikatów, numery kolejne nadawane przez serwer oraz
// opóźnione zatwierdzenie transakcji (wiersz z niższym numerem widoczny później — test zapasu kursora).
// Dane w pamięci; konta i hasła fikcyjne.
export function fakeSupabase({ anonKey = 'anon-test-key', tokenTtl = 3600, clock = () => Date.now() } = {}) {
  const users = new Map();          // email -> { id, email, password }
  const access = new Map();         // token -> { userId, exp }
  const refresh = new Map();        // token -> userId
  const events = [];                // { user_id, sid, seq, blob, visible }
  const keys = new Map();           // user_id -> { salt, iterations, verifier }
  let seq = 0, n = 0, holdNext = 0;
  const log = [];                   // [method, path] — do asercji
  const tok = p => `${p}-${++n}-${Math.random().toString(36).slice(2)}`;
  const json = (status, body) => new Response(body === undefined ? '' : JSON.stringify(body), { status, headers: { 'Content-Type': 'application/json' } });
  const session = u => {
    const a = tok('at'), r = tok('rt');
    access.set(a, { userId: u.id, exp: clock() / 1000 + tokenTtl }); refresh.set(r, u.id);
    return { access_token: a, refresh_token: r, expires_in: tokenTtl, expires_at: Math.floor(clock() / 1000 + tokenTtl), user: { id: u.id, email: u.email } };
  };
  const who = h => {
    const t = (h.Authorization || '').replace(/^Bearer /, '');
    const a = access.get(t);
    return a && a.exp > clock() / 1000 ? a.userId : null;
  };

  async function fetch(url, init = {}) {
    const u = new URL(url), h = init.headers || {}, method = init.method || 'GET';
    log.push([method, u.pathname]);
    if (h.apikey !== anonKey) return json(401, { message: 'Invalid API key' });
    const body = init.body ? JSON.parse(init.body) : null;
    // ---------- Auth
    if (u.pathname === '/auth/v1/token') {
      const grant = u.searchParams.get('grant_type');
      if (grant === 'password') {
        const usr = users.get(body?.email);
        return usr && usr.password === body.password ? json(200, session(usr)) : json(400, { error: 'invalid_grant', error_description: 'Invalid login credentials' });
      }
      if (grant === 'refresh_token') {
        const uid = refresh.get(body?.refresh_token);
        if (!uid) return json(400, { error: 'invalid_grant', error_description: 'Invalid Refresh Token' });
        refresh.delete(body.refresh_token);                                        // rotacja: stary token jednorazowy
        return json(200, session([...users.values()].find(x => x.id === uid)));
      }
      return json(400, { error: 'unsupported_grant_type' });
    }
    if (u.pathname === '/auth/v1/logout') { const t = (h.Authorization || '').replace(/^Bearer /, ''); access.delete(t); return json(204); }
    // ---------- PostgREST z RLS
    const uid = who(h);
    if (!uid) return json(401, { message: 'JWT expired or missing', code: 'PGRST301' });
    if (u.pathname === '/rest/v1/sync_keys') {
      if (method === 'GET') return json(200, keys.has(uid) ? [keys.get(uid)] : []);
      if (method === 'POST') {
        if (body.user_id !== uid) return json(403, { message: 'new row violates row-level security policy' });
        if (keys.has(uid)) return json(409, { message: 'duplicate key value violates unique constraint' });
        keys.set(uid, { salt: body.salt, iterations: body.iterations, verifier: body.verifier }); return json(201);
      }
      if (method === 'DELETE') { keys.delete(uid); return json(204); }
    }
    if (u.pathname === '/rest/v1/events') {
      if (method === 'GET') {
        const gt = Number((u.searchParams.get('seq') || 'gt.0').replace('gt.', ''));
        const limit = Number(u.searchParams.get('limit') || 1000);
        const rows = events.filter(e => e.user_id === uid && e.visible && e.seq > gt).sort((a, b) => a.seq - b.seq).slice(0, limit);
        return json(200, rows.map(e => ({ sid: e.sid, seq: e.seq, blob: e.blob })));
      }
      if (method === 'POST') {
        const ignore = /resolution=ignore-duplicates/.test(h.Prefer || '');
        if (body.some(r => r.user_id !== uid)) return json(403, { message: 'new row violates row-level security policy' });
        const dup = body.filter(r => events.some(e => e.user_id === uid && e.sid === r.sid));
        if (dup.length && !ignore) return json(409, { message: 'duplicate key value violates unique constraint' });
        const hold = holdNext > 0;
        for (const r of body) if (!events.some(e => e.user_id === uid && e.sid === r.sid)) events.push({ user_id: uid, sid: r.sid, blob: r.blob, seq: ++seq, visible: !hold });
        if (hold) holdNext--;
        return json(201);
      }
      if (method === 'DELETE') { for (let i = events.length - 1; i >= 0; i--) if (events[i].user_id === uid) events.splice(i, 1); return json(204); }
    }
    return json(404, { message: 'not found' });
  }

  return {
    fetch, anonKey, url: 'http://localhost:54321', log, events, keys, access,
    addUser(email, password) { const usr = { id: `00000000-0000-4000-8000-${String(users.size + 1).padStart(12, '0')}`, email, password }; users.set(email, usr); return usr; },
    // Następne wstawienie dostaje numery kolejne, ale jest niewidoczne do release() — jak transakcja zatwierdzona później
    holdNextInsert() { holdNext++; },
    release() { events.forEach(e => { e.visible = true; }); },
    expireAll() { for (const a of access.values()) a.exp = 0; },
  };
}

// Pamięć sesji zgodna z interfejsem `storage` klienta (w aplikacji: magazyn `meta` IndexedDB)
export const memoryStorage = () => { const m = new Map(); return { get: async k => m.get(k) ?? null, set: async (k, v) => { m.set(k, structuredClone(v)); }, remove: async k => { m.delete(k); } }; };

File: tests/e2e/fake_supabase.py
"""Lokalny FAŁSZYWY serwer zgodny z używanym podzbiorem REST Supabase (Auth + PostgREST) — wyłącznie do testów w przeglądarce.
Odpowiednik tests/unit/helpers/fake-supabase.mjs: logowanie hasłem, odświeżanie z rotacją, wygasanie tokenów, RLS (każdy
użytkownik widzi tylko swoje wiersze, bez tokenu brak dostępu), wstawianie z pominięciem duplikatów, numery nadawane przez
serwer, CORS (aplikacja działa pod innym adresem niż serwer, wariant jednoplikowy — z file://). Dane w pamięci; konta fikcyjne."""
import json, threading, time, uuid, http.server
from urllib.parse import urlsplit, parse_qs


class FakeSupabase:
    def __init__(self, anon_key, port, token_ttl=3600):
        self.anon_key, self.port, self.ttl = anon_key, port, token_ttl
        self.users, self.access, self.refresh = {}, {}, {}
        self.events, self.keys, self.log = [], {}, []
        self.seq, self.lock = 0, threading.Lock()
        self.stats = {'bytes_out': 0, 'gets': 0, 'posts': 0}   # transfer z serwera (GET /events) i liczba zapytań
        self.calls = []   # (user-agent, metoda, ścieżka, zapytanie, bajty treści) — rozróżnienie profili testowych
        self.url = f'http://localhost:{port}'

    def add_user(self, email, password):
        self.users[email] = {'id': str(uuid.UUID(int=len(self.users) + 1)), 'email': email, 'password': password}

    def expire_all(self):
        with self.lock:
            for a in self.access.values(): a['exp'] = 0

    def _session(self, u):
        a, r = 'at-' + uuid.uuid4().hex, 'rt-' + uuid.uuid4().hex
        self.access[a] = {'uid': u['id'], 'exp': time.time() + self.ttl}; self.refresh[r] = u['id']
        return {'access_token': a, 'refresh_token': r, 'expires_in': self.ttl, 'expires_at': int(time.time() + self.ttl),
                'user': {'id': u['id'], 'email': u['email']}}

    def handle(self, method, path, query, headers, body):
        """Zwraca (status, obiekt JSON albo None)."""
        self.log.append((method, path))
        if headers.get('apikey') != self.anon_key: return 401, {'message': 'Invalid API key'}
        if path == '/auth/v1/token':
            grant = query.get('grant_type', [''])[0]
            if grant == 'password':
                u = self.users.get((body or {}).get('email'))
                if u and u['password'] == body.get('password'): return 200, self._session(u)
                return 400, {'code': 400, 'error_code': 'invalid_credentials', 'msg': 'Invalid login credentials'}
            if grant == 'refresh_token':
                uid = self.refresh.pop((body or {}).get('refresh_token'), None)   # rotacja: token jednorazowy
                if not uid: return 400, {'code': 400, 'error_code': 'refresh_token_not_found', 'msg': 'Invalid Refresh Token: Refresh Token Not Found'}
                return 200, self._session(next(u for u in self.users.values() if u['id'] == uid))
            return 400, {'msg': 'unsupported_grant_type'}
        token = (headers.get('authorization') or '').removeprefix('Bearer ')
        if path == '/auth/v1/logout': self.access.pop(token, None); return 204, None
        a = self.access.get(token)
        uid = a['uid'] if a and a['exp'] > time.time() else None
        if not uid: return 401, {'code': 'PGRST301', 'message': 'JWT expired'}
        if path == '/rest/v1/sync_keys':
            if method == 'GET': return 200, [self.keys[uid]] if uid in self.keys else []
            if method == 'POST':
                if body.get('user_id') != uid: return 403, {'message': 'new row violates row-level security policy'}
                if uid in self.keys: return 409, {'message': 'duplicate key value violates unique constraint'}
                self.keys[uid] = {k: body[k] for k in ('salt', 'iterations', 'verifier')}; return 201, None
            if method == 'DELETE': self.keys.pop(uid, None); return 204, None
        if path == '/rest/v1/events':
            if method == 'GET':
                f = query.get('seq', ['gt.0'])[0]; limit = int(query.get('limit', ['1000'])[0])
                cols = query.get('select', ['sid,seq,blob'])[0].split(',')
                if f.startswith('in.('): want = {int(x) for x in f[4:-1].split(',') if x}; match = lambda e: e['seq'] in want
                else: gt = int(f.removeprefix('gt.')); match = lambda e: e['seq'] > gt
                rows = sorted((e for e in self.events if e['user_id'] == uid and match(e)), key=lambda e: e['seq'])[:limit]
                out = [{c: e[c] for c in cols} for e in rows]
                self.stats['bytes_out'] += len(json.dumps(out)); self.stats['gets'] += 1
                return 200, out
            if method == 'POST':
                if any(r.get('user_id') != uid for r in body): return 403, {'message': 'new row violates row-level security policy'}
                have = {e['sid'] for e in self.events if e['user_id'] == uid}
                if any(r['sid'] in have for r in body) and 'ignore-duplicates' not in (headers.get('prefer') or ''): return 409, {'message': 'duplicate key'}
                inserted = []
                for r in body:
                    if r['sid'] not in have:
                        self.seq += 1; have.add(r['sid']); inserted.append({'seq': self.seq})
                        self.events.append({'user_id': uid, 'sid': r['sid'], 'blob': r['blob'], 'seq': self.seq})
                self.stats['posts'] += 1
                return 201, (inserted if 'return=representation' in (headers.get('prefer') or '') else None)
            if method == 'DELETE': self.events = [e for e in self.events if e['user_id'] != uid]; return 204, None
        return 404, {'message': 'not found'}

    def start(self):
        fake = self

        class H(http.server.BaseHTTPRequestHandler):
            def log_message(self, *a): pass
            def cors(self):
                self.send_header('Access-Control-Allow-Origin', '*')
                self.send_header('Access-Control-Allow-Headers', 'apikey, authorization, content-type, prefer')
                self.send_header('Access-Control-Allow-Methods', 'GET, POST, DELETE, OPTIONS')
            def do_OPTIONS(self):
                self.send_response(204); self.cors(); self.send_header('Content-Length', '0'); self.end_headers()
            def any(self):
                u = urlsplit(self.path); n = int(self.headers.get('Content-Length') or 0)
                raw = self.rfile.read(n) if n else b''
                hdrs = {k.lower(): v for k, v in self.headers.items()}
                with fake.lock:
                    status, obj = fake.handle(self.command, u.path, parse_qs(u.query), hdrs, json.loads(raw) if raw else None)
                data = b'' if obj is None else json.dumps(obj).encode()
                with fake.lock: fake.calls.append((hdrs.get('user-agent', ''), self.command, u.path, u.query, len(data)))
                self.send_response(status); self.cors()
                self.send_header('Content-Type', 'application/json'); self.send_header('Content-Length', str(len(data))); self.end_headers()
                self.wfile.write(data)
            do_GET = do_POST = do_DELETE = any

        self.srv = http.server.ThreadingHTTPServer(('127.0.0.1', self.port), H)
        threading.Thread(target=self.srv.serve_forever, daemon=True).start()
        return self

    def stop(self): self.srv.shutdown()
